text/0000-impl-trait-type-alias.md

```diff
@@ -0,0 +1,590 @@
+    type Item = impl Debug;
+    fn next(&mut self) -> Option<Self::Item> {
+        Some("Another item!")
+    }
+}
+```
+
+In this trait implementation, we've declared that the item returned by our
+iterator implements `Debug`, but we've kept its concrete type (`&'static str`)
+hidden from the outside world.
+
+We can even use this feature to specify unnameable associated types, such as
+closures:
+
+```rust
+struct MyType;
+impl Iterator for MyType {
+    type Item = impl Fn(i32) -> i32;
+    fn next(&mut self) -> Option<Self::Item> {
+        Some(|x| x + 5)
+    }
+}
+```
+
+`impl Trait` aliases can also be used to reference unnameable types in a struct
+definition:
+
+```rust
+type Foo = impl Debug;
+fn foo() -> Foo { 5i32 }
+
+struct ContainsFoo {
+    some_foo: Foo
+}
+```
+
+It's also possible to write generic `impl Trait` aliases:
+
+```rust
+#[derive(Debug)]
+struct MyStruct<T: Debug> {
+    inner: T
+};
+
+type Foo<T> = impl Debug;
+
+fn get_foo<T: Debug>(x: T) -> Foo<T> {
+    MyStruct {
+        inner: x
+    }
+}
+```
+
+`impl Trait` can also appear inside of another type in a type alias:
+
+```rust
+type Foo = Option<impl Debug>;
+fn foo() -> Foo {
+    Some("Debuggable")
+}
+```
+
+Or even multiple times within the same type alias:
+
+```rust
+type Foo = (impl Debug, impl Fn());
+fn foo() -> Foo {
+    ("Debuggable", || println!("Hello, world!"))
+}
+```
+
+# Reference-Level Explanation
+[reference]: #reference
+
+## Reference: `impl Trait` in `let`, `const` and `static`:
+[reference-declarations]: #reference-declarations
+
+The rules for `impl Trait` values in `let`, `const`, and `static` declarations
+work mostly the same as `impl Trait` return values as specified in
+[RFC 1951](https://github.com/rust-lang/rfcs/blob/master/text/1951-expand-impl-trait.md).
+
+These values hide their concrete type and can only be used as a value which
+is known to implement the specified traits. They inherit any type parameters
+in scope. One difference from `impl Trait` return types is that they also
+inherit any lifetime parameters in scope. This is necessary in order for
+`let` bindings to use `impl Trait`. `let` bindings often contain references
+which last for anonymous scope-based lifetimes, and annotating these lifetimes
+manually would be impossible.
+
+## Reference: `impl Trait` Type Aliases
+[reference-aliases]: #reference-aliases
+
+`impl Trait` type aliases are similar to normal type aliases, except that their
+concrete type is determined from the module in which they are defined.
+For example, the following code has to examine the body of `foo` in order to
+determine that the concrete type of `Foo` is `i32`:
+
+```rust
+type Foo = impl Debug;
+
+fn foo() -> Foo {
+    5i32
+}
+```
+
+`Foo` can be used as `i32` in multiple places throughout the module.
+However, each function that uses `Foo` as `i32` must independently place
+constraints upon `Foo` such that it *must* be `i32`:
+
+```rust
+fn add_to_foo_1(x: Foo) {
```

Comment:
It's not that it's treated differently in inference vs. non-inference contexts-- it's that a value of type `Foo` only needs to be constrained to `i32` if you want to use it as an `i32`. `let _: Foo;` is fine because it doesn't need to know anything about `Foo` other than that it is a type that implements `Debug`. If you wanted to, say, add Foo: `fn add(x: Foo) { x + 5 }` then you'd need to specify that `Foo` is an `i32` by writing something like `fn add(x: Foo) { let x: i32 = x; x + 5 }`